### PR TITLE
UI-SongsCollectionView

### PR DESCRIPTION
**Pull request by at-vytran**

### CODE REVIEW

File: MyApp/DataManagement.swift [hunk 1]
```diff
@@ -16,16 +16,23 @@ final class DataManagement {
...
         var songs: [Song] = []
         let images: [String] = ["img_song_1", "img_song_2", "img_song_3", "img_song_4", "img_song_5", "img_song_6"]
         for i in 0 ..< images.count {
             let songName = "Song \(i)"
             let artistName = "Artist \(i)"
-            let song = Song(songName: songName, artistName: artistName, avatar: images[i])
+            let song = Song(songName: songName, artistName: artistName, songImage: images[i])
             songs.append(song)
         }
         return songs
```
Comment (at-daiho) [inline MyApp/DataManagement.swift]: WARNING - Lines should not have trailing whitespace.

File: MyApp/DataManagement.swift [hunk 2]
```diff
@@ -9,21 +9,24 @@
...
-    static var shared: DataManagement = {
+    public static var shared: DataManagement = {
         var dataManagement = DataManagement()
         return dataManagement
     }()
 
-     func fetchRecentlyPlayedSongs() -> [Song] {
-        // TODO: - Dummy data
+    var recentlyPlayedSongs: [Song] = {
+        return fetchRecentlyPlayedSongs()
```
Comment (at-daiho) [inline MyApp/DataManagement.swift]: WARNING - Lines should not have trailing whitespace.

File: MyApp/AppDelegate.swift [hunk 3]
```diff
@@ -82,6 +82,16 @@ final class AppDelegate: UIResponder, UIApplicationDelegate {
         textFieldInsideSearchBar.textColor = App.Color.searchBarTextColor
         view.navigationItem.titleView = searchBar
     }
+
+    func getVisibleViewHeight() -> CGFloat {
+        let statusBar = App.Measurement.statusBarHeight
+        let navigationBar = App.Measurement.navigationBarHeight
+        let topBar = statusBar + navigationBar
+        let bottomBar = App.Measurement.tabBarHeight
+
```
Comment (at-daiho) [inline MyApp/AppDelegate.swift]: WARNING - Lines should not have trailing whitespace.

File: MyApp/View/Controller/ActionViewController/Cell/Action/ActionCell.swift [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import UIKit
+
+final class ActionCell: UITableViewCell {
+
+    // MARK: - outlet
+    @IBOutlet private weak var actionTitleLabel: UILabel!
+    @IBOutlet private weak var actionImageView: UIImageView!
+
+    // MARK: - properties
+    var viewModel:  ActionCellViewModel? {
```
Comment (at-daiho) [inline MyApp/View/Controller/ActionViewController/Cell/Action/ActionCell.swift]: WARNING - Colons should be next to the identifier when specifying a type and next to the key in dictionary literals.

File: MyApp/View/Controller/ActionViewController/Cell/Close/CloseCell.swift [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+//  Created by Tran Thuy Vy on 8/7/18.
+//  Copyright © 2018 Asian Tech Co., Ltd. All rights reserved.
+//
+
+import UIKit
+
+class CloseCell: UITableViewCell {
+    
+    // MARK: - properties
+    var closeButtonTapped : (() -> Void)? = nil
```
Comment (at-daiho) [inline MyApp/View/Controller/ActionViewController/Cell/Close/CloseCell.swift]: WARNING - Initializing an optional variable with nil is redundant.

File: MyApp/View/Controller/ActionViewController/Cell/Info/InfoCell.swift [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+final class InfoCell: UITableViewCell {
+
+    // MARK: - outlet
+    @IBOutlet private weak var artistTitle: UILabel!
+    @IBOutlet private weak var songTitle: UILabel!
+    @IBOutlet private weak var songImageView: UIImageView!
+
+    // MARK: - properties
+    var viewModel: InfoCellViewModel? {
+        didSet{
```
Comment (at-daiho) [inline MyApp/View/Controller/ActionViewController/Cell/Info/InfoCell.swift]: WARNING - Opening braces should be preceded by a single space and on the same line as the declaration.

**Comment by at-daiho** [comment]
Total linter violations in pull request: 25
Serious: 0
Warnings: 25

File: MyApp/Model/Schema/Song.swift [hunk 7]
```diff
@@ -10,13 +10,13 @@ import Foundation
 import UIKit
 
 class Song {
-    var songName: String
-    var artistName: String
-    var avatar: String
+    var name: String
+    var artist: String
+    var image: String
```
Comment (at-daiho) [inline MyApp/Model/Schema/Song.swift]: coverPath

File: MyApp/Model/Schema/Song.swift [hunk 8]
```diff
@@ -10,13 +10,13 @@ import Foundation
...
 
-    init(songName: String, artistName: String, avatar: String) {
-        self.songName = songName
-        self.artistName = artistName
-        self.avatar = avatar
+    init(songName: String, artistName: String, songImage: String) {
```
Comment (at-daiho) [inline MyApp/Model/Schema/Song.swift]: init(name: String, artist: String, coverPath: String)

File: MyApp/View/Controller/SongsViewController/SongsViewModel.swift [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    }
+}
+
+extension SongsViewModel {
+    func numberOfItems(inSection section: Int) -> Int {
+        return numberOfSong
+    }
+
+    func viewModelForCell(at indexPath: IndexPath) -> SongsCellViewModel {
+        if indexPath.row <= numberOfSong - 1 {
```
Comment (at-daiho) [inline MyApp/View/Controller/SongsViewController/SongsViewModel.swift]: indexPath.row <= numberOfSong - 1 

Có tương đương với 
indexPath.row < numberOfSong ko?

File: MyApp/DataManagement.swift (hunk 1 above)
Comment (at-vytran) [reply MyApp/DataManagement.swift]: đã fix

File: MyApp/AppDelegate.swift (hunk 3 above)
Comment (at-vytran) [reply MyApp/AppDelegate.swift]: đã fix

File: MyApp/View/Controller/ActionViewController/Cell/Close/CloseCell.swift (hunk 5 above)
Comment (at-vytran) [reply MyApp/View/Controller/ActionViewController/Cell/Close/CloseCell.swift]: đã fix

File: MyApp/View/Controller/ActionViewController/Cell/Action/ActionCell.swift (hunk 4 above)
Comment (at-vytran) [reply MyApp/View/Controller/ActionViewController/Cell/Action/ActionCell.swift]: đã fix

File: MyApp/View/Controller/ActionViewController/Cell/Info/InfoCell.swift (hunk 6 above)
Comment (at-vytran) [reply MyApp/View/Controller/ActionViewController/Cell/Info/InfoCell.swift]: đã fix

File: MyApp/Model/Schema/Song.swift (hunk 7 above)
Comment (at-vytran) [reply MyApp/Model/Schema/Song.swift]: đã fix

File: MyApp/Model/Schema/Song.swift (hunk 8 above)
Comment (at-vytran) [reply MyApp/Model/Schema/Song.swift]: đã fix

File: MyApp/View/Controller/SongsViewController/SongsViewModel.swift (hunk 9 above)
Comment (at-vytran) [reply MyApp/View/Controller/SongsViewController/SongsViewModel.swift]: đã fix

**Comment by at-daiho** [comment]
:+1: There are no linter violations.

**Comment by at-daiho** [comment]
Total linter violations in pull request: 4
Serious: 0
Warnings: 4

**Comment by at-daiho** [comment]
Total linter violations in pull request: 3
Serious: 0
Warnings: 3

File: MyApp/View/Controller/ActionViewController/Cell/ActionViewController.swift [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    func showActionSheet() {
+        UIView.animate(withDuration: 0.3, animations: {
+            self.tableView.frame.origin.y = self.view.frame.size.height - 250
+        })
+    }
+
+    private func hideActionSheet() {
+        UIView.animate(withDuration: 0.3, animations: {
+            self.tableView.frame.origin.y = self.view.frame.size.height
+        }) { (done) in
```
Comment (at-daiho) [inline MyApp/View/Controller/ActionViewController/Cell/ActionViewController.swift]: WARNING - Trailing closure syntax should not be used when passing more than one closure argument.

**Comment by at-daiho** [comment]
Total linter violations in pull request: 1
Serious: 0
Warnings: 1

**Pull request merged**